src/com/xilinx/rapidwright/design/NetTools.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2024, Advanced Micro Devices, Inc.
+ * All rights reserved.
+ *
+ * Author: Wenhao Lin, Advanced Micro Devices, Inc.
+ *
+ * This file is part of RapidWright.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.xilinx.rapidwright.design;
+
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Set;
+
+import com.xilinx.rapidwright.device.SiteTypeEnum;
+
+public class NetTools {
+    private static Set<SiteTypeEnum> clkSrcSiteTypeEnums = EnumSet.noneOf(SiteTypeEnum.class);
+    static {
+        clkSrcSiteTypeEnums.addAll(List.of(
+            SiteTypeEnum.BUFGCE,        // All supported series
+            SiteTypeEnum.BUFGCTRL,      // All supported series
+            SiteTypeEnum.BUFG,          // All supported series
+            SiteTypeEnum.BUFGCE_DIV,    // US/US+ and Versal
+            SiteTypeEnum.BUFG_GT,       // US/US+ and Versal
+            SiteTypeEnum.BUFG_PS,       // US/US+ and Versal
+            SiteTypeEnum.BUFG_FABRIC,   // Versal
+            SiteTypeEnum.BUFIO,         // Series-7
+            SiteTypeEnum.BUFR,          // Series-7
+            SiteTypeEnum.BUFHCE,        // Series-7
```

Comment:
> HROW Clock Buffer for a Single Clocking Region with Clock Enable

https://docs.amd.com/r/en-US/ug953-vivado-7series-libraries/BUFHCE

Please remove.

Reply:
This may be needed.
Codes:
```
       for (Net net: design.getNets()) {
            if (NetTools.isGlobalClock(net)) {
                globalClockNamesFromNetTools.add(net.getName());
                if (design.getDevice().getSeries() == Series.Series7) {
                    System.out.println(net.getName() + " " + net.getSource().getSiteTypeEnum());
                }
            }
        }
```
Output:

Series: Series7
CLK_BUFG_BOT_R_X60Y48_BUFGCTRL_X0Y0_O BUFHCE
-----------Vivado------------
CLK_BUFG_BOT_R_X60Y48_BUFGCTRL_X0Y0_O
-----------NetTools-----------
CLK_BUFG_BOT_R_X60Y48_BUFGCTRL_X0Y0_O

Reply:
The above test design is bug349.dcp.

Reply:
Double check Vivado output:
![image](https://github.com/user-attachments/assets/de4f3d54-ef08-45df-b04b-47e05744a618)

Reply:
On `bug349.dcp` I'm seeing `BUFGCTRL` not `BUFHCE`?

Reply:
> On `bug349.dcp` I'm seeing `BUFGCTRL` not `BUFHCE`?
In RWRoute I saw the source `SitePinInst` of this net is BUFHCE_X0Y11/O with `BUFHCE` type.

In Vivado I saw the source cell pin of this net is
![image](https://github.com/user-attachments/assets/f20ad40e-f5ea-487d-be6e-182d61faddb6)
![image](https://github.com/user-attachments/assets/bd2b37f6-a498-4275-8dd7-5d1dab143df0)
which is aligned to your seeing. 
Then, I looked for the node of the source `SitePinInst` BUFHCE_X0Y11/O reported by RWRoute, and found CLK_HROW_BOT_R_X60Y26/CLK_HROW_CK_HCLK_OUT_L11
![image](https://github.com/user-attachments/assets/7ef3abd1-beee-4430-b05c-b75a20c7e6a9)
is also in the node list of the clock net.

Reply:
> is also in the node list of the clock net.

Correct -- it is a node in the clock net, but not the source node/pin:
```
report_route_status -of [get_nets CLK_BUFG_BOT_R_X60Y48_BUFGCTRL_X0Y0_O]
============================================================
Route information for CLK_BUFG_BOT_R_X60Y48_BUFGCTRL_X0Y0_O
  Route status: ROUTED
  This net is fully routed, and has locked routing nodes
------------------------------------------------------------
  The route tree for this net is:
    Route Tree: 
    ------------------------------------------ 
    Subtree: 0
    [{*      CLK_BUFG_BOT_R_X60Y48/CLK_BUFG_BUFGCTRL0_O (65535) 
      *      CLK_BUFG_REBUF_X60Y65/CLK_BUFG_REBUF_R_CK_GCLK0_BOT ( 1) CLK_BUFG_BOT_R_X60Y48/CLK_BUFG_BOT_R.CLK_BUFG_BUFGCTRL0_O->>CLK_BUFG_CK_GCLK0
      *      CLK_BUFG_REBUF_X60Y38/CLK_BUFG_REBUF_R_CK_GCLK0_BOT ( 0) CLK_BUFG_REBUF_X60Y38/CLK_BUFG_REBUF.CLK_BUFG_REBUF_R_CK_GCLK0_BOT<<->>CLK_BUFG_REBUF_R_CK_GCLK0_TOP
      *      CLK_HROW_BOT_R_X60Y26/CLK_HROW_CK_MUX_OUT_L11 (30) CLK_HROW_BOT_R_X60Y26/CLK_HROW_BOT_R.CLK_HROW_R_CK_GCLK0->>CLK_HROW_CK_MUX_OUT_L11
      *      CLK_HROW_BOT_R_X60Y26/CLK_HROW_CK_HCLK_OUT_L11 ( 0) CLK_HROW_BOT_R_X60Y26/CLK_HROW_BOT_R.CLK_HROW_CK_MUX_OUT_L11->>CLK_HROW_CK_HCLK_OUT_L11
              RouteThru, site: BUFHCE_X0Y11 From: I To: O 
      *      CLK_HROW_BOT_R_X60Y26/CLK_HROW_CK_BUFHCLK_L11 ( 0) CLK_HROW_BOT_R_X60Y26/CLK_HROW_BOT_R.CLK_HROW_CK_HCLK_OUT_L11->>CLK_HROW_CK_BUFHCLK_L11
      *      HCLK_L_X15Y26/HCLK_LEAF_CLK_B_BOTL4 ( 3) HCLK_L_X15Y26/HCLK_L.HCLK_CK_BUFHCLK11->>HCLK_LEAF_CLK_B_BOTL4
     {*         INT_L_X4Y13/GCLK_L_B10_WEST ( 0) INT_L_X4Y13/INT_L.GCLK_L_B10->>GCLK_L_B10_WEST
      *                  INT_L_X4Y13/CLK_L1 ( 4) INT_L_X4Y13/INT_L.GCLK_L_B10_WEST->>CLK_L1
      *  }       CLBLL_L_X4Y13/CLBLL_LL_CLK ( 0) CLBLL_L_X4Y13/CLBLL_L.CLBLL_CLK1->CLBLL_LL_CLK
      *         INT_L_X4Y12/GCLK_L_B10_WEST ( 0) INT_L_X4Y12/INT_L.GCLK_L_B10->>GCLK_L_B10_WEST
      *                  INT_L_X4Y12/CLK_L1 ( 4) INT_L_X4Y12/INT_L.GCLK_L_B10_WEST->>CLK_L1
      *  }]      CLBLL_L_X4Y12/CLBLL_LL_CLK ( 0) CLBLL_L_X4Y12/CLBLL_L.CLBLL_CLK1->CLBLL_LL_CLK
    ------------------------------------------ 
============================================================
```

Specifically, what you're seeing here is that the global net does a transparent "routethru" of the BUFHCE, but ultimately the logical source of the net is the BUFGCTRL.


With that in mind, it does look like you might have discovered a bug in RapidWright, thanks! The clock net in question is actually given two sources by RapidWright -- the BUFHCE as the primary source, and the BUFGCTRL as the alternate source. This is not correct, and requires fixing first.

Reply:
Actually, this is already fixed in an upcoming release. Currently failing test added in https://github.com/Xilinx/RapidWright/pull/1054/commits/4c29c43ec92ff5e9127428e08262ba9a2af8ac7e and attached to https://github.com/Xilinx/RapidWright/pull/1054